include/unifex/detail/debug_async_scope.hpp

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License Version 2.0 with LLVM Exceptions
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *   https://llvm.org/LICENSE.txt
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/detail/intrusive_list.hpp>
+#if __has_include(<cxxabi.h>)
+#  define UNIFEX_NO_DEMANGLE 0
+#  include <cxxabi.h>
+#else
+// TODO
+// https://learn.microsoft.com/en-us/windows/win32/api/dbghelp/nf-dbghelp-undecoratesymbolname?redirectedfrom=MSDN
+#  define UNIFEX_NO_DEMANGLE 1
+#endif
+#include <mutex>
+#include <typeinfo>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex::detail {
+namespace _debug_async_scope {
+
+struct op_base {
+  explicit op_base(const std::type_info& concreteType) noexcept
+    : concreteType(concreteType)
+    , demangled(demangle(concreteType)) {}
+
+  const std::type_info& concreteType;
+  const char* demangled;
+  bool shouldFree{false};
+
+  const char* demangle(const std::type_info& concreteType) noexcept {
+#if !UNIFEX_NO_DEMANGLE
+    int status = -1;
+    auto result = abi::__cxa_demangle(concreteType.name(), nullptr, 0, &status);
+    if (status == 0) {
+      shouldFree = true;
+    } else {
+      return concreteType.name();
+    }
+#else
+    auto result = concreteType.name();
+#endif
+    return result;
+  }
+
+  op_base* next{nullptr};
+  op_base* prev{nullptr};
+
+  ~op_base() {
+    if (shouldFree) {
+      std::free((void*)demangled);
+    }
+  }
+};
+
+using debug_ops_t = intrusive_list<op_base, &op_base::next, &op_base::prev>;
+
+struct debug_op_list final {
+  std::mutex mutex_;
+  debug_ops_t ops_;
+
+  void register_debug_operation(op_base* op) noexcept {
+    std::lock_guard lock{mutex_};
+    ops_.push_back(op);
+  }
+
+  void deregister_debug_operation(op_base* op) noexcept {
+    std::lock_guard lock{mutex_};
+    ops_.remove(op);
+  }
+};
+
+template <typename Receiver>
+struct _operation final {
+  struct type;
+};
+
+template <typename Receiver>
+struct _operation<Receiver>::type : op_base {
+  template <typename Receiver2>
+  type(
```

Comment:
Nit: most constructors in Unifex are `explicit`.
```suggestion
  explicit type(
```